### PR TITLE
Clarifications and small fixes to to-device related code

### PR DESCRIPTION
**Pull request by anoadragon453**
A lot of little clarifications and some fixes split out from #11215 to help reduce muddying of the new feature.

Most of the comment changes are just helping document what's already there. There are also a couple code changes included:

* Tightening up a type hint that didn't need to be optional.
* Preventing shadowing the `type` built-in by renaming a variable.

Recommended to review commit-by-commit.

### CODE REVIEW

File: synapse/handlers/devicemessage.py [hunk 1]
```diff
@@ -135,12 +142,16 @@ async def on_direct_to_device_edu(self, origin: str, content: JsonDict) -> None:
                 message_type, sender_user_id, by_device
             )
 
-        stream_id = await self.store.add_messages_from_remote_to_device_inbox(
+        # Add messages to the database.
+        # Retrieve the stream id of the last-processed to-device message.
+        max_stream_id = await self.store.add_messages_from_remote_to_device_inbox(
```
Comment (DMRobertson) [inline synapse/handlers/devicemessage.py:147]: What does the `max` indicate here? Largest stream id that has been processed?

File: synapse/handlers/appservice.py [hunk 2]
```diff
@@ -203,7 +203,7 @@ def notify_interested_services_ephemeral(
                 value for `stream_key` will cause this function to return early.
 
                 Ephemeral events will only be pushed to appservices that have opted into
-                them.
+                them via the ephemeral AS registration file option.
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply synapse/handlers/appservice.py:206]: "(ephemeral AS registration file) option" is right.

Application services register themselves with a homeserver through a [registration file](https://matrix.org/docs/spec/application_service/r0.1.2#registration). That file is filled with options, for which sending ephemeral events is an opt-in one (https://github.com/Sorunome/matrix-doc/blob/soru%2Bhs/appservice-edus/proposals/2409-appservice-edus.md#changes-to-the-registration-file).

I did feel this was a little convoluted when writing it, so thanks for pointing it out. What do you think about:

> Ephemeral events will only be pushed to appservices that have opted into receiving them by setting `push_ephemeral` to true in their registration file. Note that while MSC2409 is experimental, this option is called `de.sorunome.msc2409.push_ephemeral`.

? Bit wordy but comments are free!

File: synapse/handlers/devicemessage.py (hunk 1 above)
Comment (anoadragon453) [reply synapse/handlers/devicemessage.py:147]: Yes, it's just the stream id of the last to-device message that was added to the database. Messages are added to the database in with incrementing stream ids.

```
# Retrieve the stream id of the to-device message that was stored last (the largest stream id).
```

instead perhaps?

File: synapse/handlers/appservice.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/handlers/appservice.py:206]: Bit wordy, but not that big and it's much clearer to me. :+1:

File: synapse/handlers/devicemessage.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/handlers/devicemessage.py:147]: I guess my concern here was that "processing" could be a separate step to persisting them in the database. I also wasn't sure if we might be adding messages with stream ids say `10, 11, 12`, but someone else has already put a message with a stream id of `30` into the db.

Comment (anoadragon453) [reply synapse/handlers/devicemessage.py:147]: The stream id that is assigned to each to-device message is implemented using an ever-incrementing database sequence:

https://github.com/matrix-org/synapse/blob/4535532526581834ab798996ffe73f6d19c25123/synapse/storage/databases/main/deviceinbox.py#L468-L475

So having multiple rows with the same stream id should never occur.

to-device messages are stored here. Later on they are retrieved from the database and sent off to remote homeservers/down user sync streams. Each remote has a last-known-received stream id, and the same is encoded in client sync tokens.

...does that quell your concern?

File: synapse/handlers/appservice.py [hunk 3]
```diff
@@ -252,7 +262,7 @@ async def _notify_interested_services_ephemeral(
         new_token: int,
         users: Collection[Union[str, UserID]],
     ) -> None:
-        logger.debug("Checking interested services for %s" % (stream_key))
+        logger.debug("Checking interested services for %s" % stream_key)
```
Comment (clokep) [inline synapse/handlers/appservice.py:265]: I think you wanted this?

```suggestion
        logger.debug("Checking interested services for %s", stream_key)
```

File: synapse/handlers/devicemessage.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/devicemessage.py:147]: `last_stream_id` might be a better name in this case?

File: synapse/storage/databases/main/deviceinbox.py [hunk 4]
```diff
@@ -153,12 +156,21 @@ def get_new_messages_for_device_txn(txn):
             txn.execute(
                 sql, (user_id, device_id, last_stream_id, current_stream_id, limit)
             )
+
             messages = []
+            stream_pos = current_stream_id
```
Comment (clokep) [inline synapse/storage/databases/main/deviceinbox.py:161]: Why is this added? In case `txn` is empty?

File: synapse/storage/databases/main/deviceinbox.py [hunk 5]
```diff
@@ -153,12 +156,21 @@ def get_new_messages_for_device_txn(txn):
...
+
             for row in txn:
                 stream_pos = row[0]
                 messages.append(db_to_json(row[1]))
+
+            # If we don't end up hitting the limit, we still want to return the equivalent
+            # value of `current_stream_id` to the calling function. This is needed as we'll
+            # be processing up to `current_stream_id`, without necessarily fetching a message
+            # with a stream token of `current_stream_id`.
```
Comment (clokep) [inline synapse/storage/databases/main/deviceinbox.py:170]: I'm not really following this, wouldn't this cause us to "skip" rows or something?

File: synapse/storage/databases/main/deviceinbox.py [hunk 6]
```diff
@@ -261,12 +273,19 @@ def get_new_messages_for_remote_destination_txn(txn):
             )
             txn.execute(sql, (destination, last_stream_id, current_stream_id, limit))
             messages = []
```
Comment (clokep) [inline synapse/storage/databases/main/deviceinbox.py:275]: Do we need to instantiate `stream_pos` like we do above?

It's a shame that this code looks identical minus the table name.

**Review by clokep — APPROVED** [review]

File: synapse/handlers/appservice.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/handlers/appservice.py:265]: Oh! Yes, that's a better shout.

File: synapse/storage/databases/main/deviceinbox.py (hunk 4 above)
Comment (anoadragon453) [reply synapse/storage/databases/main/deviceinbox.py:161]: It's mostly just to stop my editor complaining that `stream_pos` is never set, though we should never get into a state where `stream_pos` is `None`. We'll either pull something from the database, or we won't have and then `len(messages) < limit`...

Well, unless `limit` is 0, but that'd be unlikely to be set.

So yes, just a peace of mind safeguard really. I can add a comment above it explaining what we're doing with `stream_pos` overall.

File: synapse/storage/databases/main/deviceinbox.py (hunk 6 above)
Comment (anoadragon453) [reply synapse/storage/databases/main/deviceinbox.py:275]: Oh! I didn't notice that until now. Refactoring it is definitely an option.

File: synapse/storage/databases/main/deviceinbox.py (hunk 5 above)
Comment (anoadragon453) [reply synapse/storage/databases/main/deviceinbox.py:170]: We shouldn't, though I see that the explanation isn't clear.

For example, let's say `user_id` is `@alice:localhost`, `last_stream_id` is 1, and `current_stream_id` is 5, limit is 3, and this is what the DB looks like:

```
      user_id      |                  device_id                  | stream_id |     message_json   | instance_name 
-------------------+---------------------------------------------+-----------+--------------------+---------------
 @alice:localhost  | HAKDUDVZXZ                                  |         1 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         2 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         3 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         4 | {"content":{"..."} | master
 @bob:localhost    | TYCGLKNPVJ                                  |         5 | {"content":{"..."} | master
(5 rows)
```

We're going to pull out to-device messages with stream ids 2, 3, 4. `len(messages) == limit`, so `stream_pos` is 4. We return `4` alongside the messages, and `4` is stored.

Later on, a new to-device message comes in, so the DB now looks like:

```
      user_id      |                  device_id                  | stream_id |     message_json   | instance_name 
-------------------+---------------------------------------------+-----------+--------------------+---------------
 @alice:localhost  | HAKDUDVZXZ                                  |         1 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         2 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         3 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         4 | {"content":{"..."} | master
 @bob:localhost    | TYCGLKNPVJ                                  |         5 | {"content":{"..."} | master
 @bob:localhost    | TYCGLKNPVJ                                  |         6 | {"content":{"..."} | master
(6 rows)
```

And we call this function with `@alice:localhost`, `last_stream_id` as 4, and `current_stream_id` as 6. Limit is still 3.

We end up not pulling out any messages, as none of them are addressed to `@alice:localhost`. However, we still need to return `6` for `stream_pos`, so that the calling function knows where to check from on the next round. If we returned `4`, then we'd duplicate work checking stream ids 5, 6 again.

---

Thinking more about it now... there is an edge case where we end up returning exactly `<limit>` rows. Say you had the below DB table: 

```
      user_id      |                  device_id                  | stream_id |     message_json   | instance_name 
-------------------+---------------------------------------------+-----------+--------------------+---------------
 @alice:localhost  | HAKDUDVZXZ                                  |         1 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         2 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         3 | {"content":{"..."} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         4 | {"content":{"..."} | master
 @bob:localhost    | TYCGLKNPVJ                                  |         5 | {"content":{"..."} | master
(5 rows)
```

And you called `get_new_messages_for_device` with user_id `@alice:localhost`, `last_stream_id` as 1, `current_stream_id` as 5, with limit 3.

The function would check over rows 2 - 5 in the DB, and return rows 2, 3, 4. `len(messages) == limit`, so `stream_pos` is that of the last row we pulled out: 4. However, the function did check row 5 already. Next time round the function will be called with `last_stream_id` as 4. So we'll end up doing some duplicate work checking row 5 again.

I think this only causes a slight drop in perf though, rather than going the other way and causing us to miss rows...

File: synapse/storage/databases/main/deviceinbox.py (hunk 6 above)
Comment (clokep) [reply synapse/storage/databases/main/deviceinbox.py:275]: Yep, refactoring is more of a long term thing, just noticed it while reviewing. (I didn't closely inspect if they *are* identical by the way. They might just be very similar.)

File: changelog.d/11247.misc [hunk 7]
```diff
@@ -0,0 +1 @@
+Clean up some areas of the codebase relating to to-device messages and sending ephemeral events to application services.
```
Comment (clokep) [inline changelog.d/11247.misc:1]:
```suggestion
Clean up code relating to to-device messages and sending ephemeral events to application services.
```

File: synapse/storage/databases/main/deviceinbox.py (hunk 5 above)
Comment (clokep) [reply synapse/storage/databases/main/deviceinbox.py:170]: After chatting about this it seems that the idea is that if we don't hit the limit that means there's no data left (up to the `current_stream_id`) that relates to the requested filtering (by user/device). Thus it is safe to consider up to `current_stream_id` processed (instead considering only up to the last result's stream position).

I think the following simpler comment describes this?

```suggestion
            # If the limit was not reached we know that there's no more data for this
            # user/device pair up to current_stream_id.
```

Comment (anoadragon453) [reply synapse/storage/databases/main/deviceinbox.py:170]: I'd say that's accurate!

Comment (anoadragon453) [reply synapse/storage/databases/main/deviceinbox.py]: Fixed in a25c8dd4c55f79a20b5046c32329a4e27af705aa.

**Review by clokep — APPROVED** [review]
